fhir-persistence-blob-app/src/main/java/com/ibm/fhir/persistence/blob/app/PayloadReconciliation.java

```diff
@@ -0,0 +1,186 @@
+    
+    // Simple map cache of resource type id to name
+    private final Map<Integer,String> resourceTypeMap = new HashMap<>();
+
+    // Stop scanning after we hit this number of seconds
+    private final int maxScanSeconds;
+    /**
+     * Public constructor
+     * @param tenantId
+     * @param dsId
+     * @param dbProperties
+     * @param dbType
+     * @param dryRun
+     */
+    public PayloadReconciliation(String tenantId, String dsId, Properties dbProperties, DbType dbType, boolean dryRun, int maxScanSeconds) {
+        this.tenantId = tenantId;
+        this.dsId = dsId;
+        this.dryRun = dryRun;
+        this.maxScanSeconds = maxScanSeconds;
+
+        this.dbSupport = new DatabaseSupport(dbProperties, dbType);
+        this.dbSupport.init();
+    }
+
+    /**
+     * Run the reconciliation process
+     * @return the continuation token used to start scanning from where we left off
+     */
+    public String run(String continuationToken) throws Exception {
+        long start = System.nanoTime();
+        // Set up the request context for the configured tenant and datastore
+        FHIRRequestContext.set(new FHIRRequestContext(tenantId, dsId));
+        
+        // Keep processing until we make no more progress
+        BlobManagedContainer bmc = BlobContainerManager.getSessionForTenantDatasource();
+        BlobResourceScanner scanner = new BlobResourceScanner(continuationToken, page -> processPage(bmc, page));
+        scanner.run(bmc, maxScanSeconds);
+        
+        long end = System.nanoTime();
+        double elapsed = (end - start) / 1e9;
+        logger.info(String.format("Processed %d records in %5.1f seconds [rate %5.1f resources/second]", 
+            totalProcessed, elapsed, totalProcessed/elapsed));
+
+        return scanner.getContinuationToken();
+    }
+    
+    /**
+     * Consumer to process a page of records retrieved by the scanner
+     * @param session
+     * @param connecion
+     * @param record
+     */
+    private void processPage(BlobManagedContainer bmc, List<ResourceRecord> page) {
+
+        // Process each page in its own transaction
+        try (ITransaction tx = dbSupport.getTransaction()) {
+            try (Connection c = dbSupport.getConnection()) {
+                for (ResourceRecord record: page) {
+                    this.totalProcessed++;
+                    if (logger.isLoggable(Level.FINE)) {
+                        logger.fine(getLogRecord(record, "CHECK"));
+                    }
+                    
+                    // Check that we have the resource in the RDBMS configured for
+                    // this tenant
+                    try {
+                        ResourceExistsDAO dao = new ResourceExistsDAO(this.resourceTypeMap, 
+                            record.getResourceTypeId(), record.getLogicalId(), record.getVersion(), 
+                            record.getResourcePayloadKey());
+                        if (dao.run(c)) {
```

Comment:
i assume that deleting one-at-a-time is fine since we think this should be a very rare condition?

Reply:
I was thinking we'd do this at some point in the future, but now I'm thinking it's probably worth getting this into the next release.

Reply:
done; definitely quicker